osd: migrate from boost::variant to std::variant

tchaikov opened the pull request:
Replace boost::variant with std::variant throughout the OSD-related codebase. This change reduces third-party dependencies by leveraging the C++ standard library alternative.

Changes:
- common/inline_variant.h: Replace the existing match() helper with a wrapper around std::visit. The previous implementation constructed a visitor class from given functions; the new implementation provides equivalent functionality using standard library primitives.

- osd/osd_types.h: Add templated operator<< overload for std::variant. Since boost::variant provided a built-in operator<< that we relied on, and std::variant does not include this functionality, we implement our own formatter. To avoid ambiguous overload resolution (where types implicitly convertible to variant alternatives could match both the variant formatter and their native formatters), the template requires at least one alternative type parameter.

This migration maintains existing functionality while eliminating the boost::variant dependency.





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

ronen-fr commented on src/common/inline_variant.h:
```diff
@@ -1,211 +1,19 @@
 {
-    return boost::apply_visitor(detail::make_visitor<Variant>(
-        boost::forward<Functions>(functions)...), variant);
+    return std::visit(overloaded{std::forward<Functions>(functions)...}, variant);
```
nice!

ronen-fr commented on src/osd/PGTransaction.h:
```diff
@@ -208,12 +209,12 @@ class PGTransaction {
 	  [&](const BufferUpdate::Zero &) -> bool {
-	    auto r = boost::get<BufferUpdate::Zero>(&right);
-	    return r != nullptr;
+	    auto is_zero = std::holds_alternative<BufferUpdate::Zero>(right);
```
(taste issue) no need for the 'is_zero'. The hold_alternative is clear enough.

tchaikov replied:
i concur with you, Ronen. just wanted to keep the existing structure. will change accordingly.

ronen-fr commented on src/osd/osd_types.h:
```diff
@@ -1167,7 +1167,7 @@ class pool_opts_t {
       return false;
     }
-    *val = boost::get<T>(i->second);
+    *val = std::get<T>(i->second);
```
Just to make sure: that's definitely a get<> and not a get_if<> ?

tchaikov replied:
yes, we assign the *value* of the alternative to the pointee of `val`. so it's definitely `std::get<>`.

ronen-fr replied:
Sure. I was referring to the behaviour when 'not the current alternative' - not the taking of ref.

tchaikov replied:
I see. I intended to preserve the existing behavior. If “not the specified alternative”, an exception is thrown. but I don’t think to verify the existing behavior is in the scope of this change, what do you think?

ronen-fr replied:
I agree

ronen-fr commented:
jenkins test api

ronen-fr commented:
jenkins test make check arm64

ronen-fr reviewed (approved):
LGTM (with one question in the comments)

tchaikov commented:
changelog:

- dropped `is_zero` variable to simplify the implementation.
- updated `src/test/osd/test_pg_transaction.cc` accordingly to fix the FTBFS.

tchaikov commented:
jenkins test api

tchaikov commented:
jenkins test windows

tchaikov commented:
jenkins test windows

tchaikov commented:
jenkins test api

tchaikov commented:
jenkins test windows

ronen-fr commented:
jenkins test windows

ronen-fr commented:
jenkins test api

tchaikov commented:
jenkins test windows

tchaikov commented:
jenkins test api

tchaikov commented:
jenkins test api

tchaikov commented:
jenkins test windows

tchaikov commented:
jenkins test windows

tchaikov commented:
jenkins test api

tchaikov commented:
jenkins test make check
